modules/workflow-service-impl/src/main/java/org/opencastproject/workflow/endpoint/WorkflowRestService.java

```diff
@@ -738,18 +739,21 @@ public Response registerWorkflowDefinition(@FormParam("workflowDefinition") Work
   }
 
   @DELETE
-  @Path("/definition/{id}")
```

Comment:
Hmm.... that breaks API compatibility to existing clients... I see that point that the ID alone is not unique anymore which works out well for almost anycase except this one ;-)

Ideas for a backwards-compatible change?

Reply:
Hm, but this only breaks the internal REST API, doesn't it? The one from the remote to the implementation. Am I missing something? Are there people who update their remotes, but not their core, or vice-versa?

Reply:
The world of APIs in Opencast currently looks like:

1. The Admin UI Facade is an internal API
2. The External API is an external API
3. All other APIs are neither nor ;-)

Reply:
Maybe the following approach would make sense: DELETE /definition/{id} keeps the signature but does the following: If caller has ROLE_ADMIN, unregister all workflow definitions that have the identifier "id". If not, just unregister the workflow definition with "id" and the organization set to the current organization.

Reply:
> If caller has ROLE_ADMIN, unregister all workflow definitions (can be 0, 1 or 2) that have the identifier "id". If not, just unregister the workflow definition (can be 0 or 1) with "id" and the organization set to the current organization.

Please don't, that extremely confusing. Especially since the global admin would still access the endpoint as part of an organization.

Reply:
@lkiesow That was just an idea ;-)

With reading, there is no problem with this model because there is always exactly none or one workflow definition with a given WDID from point of view of an organization: Either none, the one specific to the organization OR the one available for all.

When updating, the question is which one to take and whether an organization is allowed to update workflow definitions that will affect other organizations. So there are two endpoints to be considered: DELETE WD & PUT WD.

As for PUT WD, I wonder whether this makes sense at all (as the change is not made persistent anyway and will therefore be lost when restarting Opencast). Actually, I wonder how much sense DELETE makes for the same reason.

Writing this, I'm tempted to propose the removal of those two REST functions as they actually don't make much sense without persistency which we don't have.

Removing those functions would result in a sound solution as all other situations work well.

What do you think? Ideas?

Reply:
> With reading, there is no problem with this model because there is always exactly none or one workflow definition with a given WDID from point of view of an organization: Either none, the one specific to the organization OR the one available for all organizations.

That's still a very weird behavior, isn't it? As a user, if I delete something, I expect it to be deleted and not replaced by another version.

If we can get rid of all the endpoints, that's, of course, a solution :D

Reply:
You cannot delete workflow definitions as a user - they can only be configured by an admin having access to Opencast's configuration files.
While there are two functions to delete or update workflow definitions using a REST API, those don't seem to make much sense in the current implementation (lack of persistence) and I personally would not aim at adding persistence as the configuration files are working just fine for at least us.

As for the configuration files, the motivation to allow "overrides" of the global workflow configuration is the assumption that most of the configuration is the same for all tenants but a provider might want to tweak individual tenants. Of course, we could just say each workflow definition has exactly one organization, that would - in our case - result in approx. 1600 workflow definition files where most of them are the same (except the organization). While Ansible can create them, I still would prefer to not see so many files in etc/workflows ...

Reply:
admins are still users ;-P

But I'm not arguing against having a tenant workflow definition. It's just that if you delete a workflow for an organization it would be very weird if it's not deleted but replaced.

But that's no problem anymore if you remove these endpoints.

Reply:
Yep, absolutely, but as an admin I would rather adjust the configuration in etc/workflows than use a REST endpoint ;-)

I'll check the state of the current implementation of the DELETE/PUT requests and bring the topic up at the tomorrows technical meeting.